jet_tracking/jet_control.py

```diff
@@ -160,49 +175,89 @@ def calibrate_inline(injector, camera, params, *, settle_time=1.0,
     burst_imagess : int, optional
         Number of burst images to average from the camera
     '''
+
     injector_axis = injector.coarseZ
+    # collect images and motor positions needed for calibration
     positions, imgs = get_calibration_images(injector_axis, camera,
                                              settle_time=settle_time,
                                              burst_images=burst_images)
 
+    # cam_roll: rotation of camera about z axis in radians
+    # pxsize: size of pixel in mm
     cam_roll, pxsize = cam_utils.get_cam_roll_pxsize(imgs, positions)
     params.pxsize.put(pxsize)
     params.cam_roll.put(cam_roll)
 
+    # beam_x_px: x-coordinate of x-ray beam in camera image in pixels
     beam_x_px = params.beam_x_px.get()
+    # beam_y_px: y-coordinate of x-ray beam in camera image in pixels
     beam_y_px = params.beam_y_px.get()
+    # cam_x: x-coordinate of camera position in mm
+    # cam_y: y-coordinate of camera position in mm
     cam_x, cam_y = cam_utils.get_cam_coords(beam_x_px, beam_y_px,
                                             cam_roll=cam_roll, pxsize=pxsize)
     params.cam_x.put(cam_x)
     params.cam_y.put(cam_y)
 
-    # find jet in camera ROI
-    ROI_image = cam_utils.get_burst_avg(burst_images, camera.ROI_image)
-    rho, theta = cam_utils.jet_detect(ROI_image)
-
-    jet_roll = cam_utils.get_jet_roll(theta, cam_roll=cam_roll)
+    # jet_roll: rotation of sample jet about z axis in radians
+    jet_roll = cam_utils.get_jet_roll(params.theta.get(), cam_roll=cam_roll)
     params.jet_roll.put(jet_roll)
-    return dict(jet_roll=jet_roll, rho=rho, theta=theta, pxsize=pxsize,
+    return dict(jet_roll=jet_roll, pxsize=pxsize,
                 cam_roll=cam_roll)
 
 
-def calibrate(injector, camera, params, *, offaxis=False, settle_time=0.1):
+def calibrate(injector, camera, cspad, wave8, gas_det, params, *, offaxis=False, settle_time=0.1):
     '''
-    Calibrate the camera
+    Calibrate the camera and CSPAD and determine parameters needed for
+    jet tracking
+    NEED TO CHECK offaxis calculation sign
+
+    First set the ROI of the camera to show the proper jet and illumination.
+
+    Determines the mean, standard deviation, radius, intensity, jet position and
+    tilt, pixel size, camera position and tilt
+
+    Params determined if onaxis camera used: mean, std, radius, intensity, pxsize,
+    camX, camY, cam_roll, jet_roll
+
+    Params determined if offaxis camera used: mean, std, radius, intensity, pxsize,
+    camY, camZ, cam_pitch, jet_pitch
 
     Parameters
     ----------
     injector : Injector
         sample injector
     camera : Questar
         camera looking at sample jet and x-rays
+    cspad : CSPAD
+        CSPAD for data
+    wave8 : Wave8
```

Comment:
Is the wave8 intensity even used here?  I only see the gasdet being used.  Looking forward, we should aim to make these calibrations more "hardware agnostic" and simply specify "upstream intensity" and "downstream intensity" readout sources.  This will remove the need for listing specific devices like this, and save us when the CSPAD is replaced by the Jungfrau or we deploy to a hutch that uses the Epix10k.  The obvious hurdle here is making these readouts look the same from an object point of view.  Maybe @silkenelson or @aegger13 can weigh in.

Reply:
I would suggest to replace 'wave8' by i0. And yes, it does get used: its a much better measure than the gas detector as it include effects of stoppers, attenuators,.... I'm not sure the gas detector should stay but we can have that discussion later. in XCS in mono, the gas detector is largely meaningless. So: there will be a minimum i0 requirement passed to the psana piece to select events. Then both the i0 and SAXS (?) sums and the shots used will be passed back via EPICS. I'm not sure SAXS is the right name: it's the liquid scattering ring intensity we are looking for using whatever detector is used for it. This will be the jungfrau4M in the main CXI chamber, the CsPad in SSC and the Epix10k2M in XCS. If there will be liquid scattering experiment where jet tracking would help in MFX in the future, we might even use the rayonix. So yes, the detector name should do. I'd have Meng and Tim come up with the suggestion and confirm that 'i0' is an acceptable generic name for the wave8.

Reply:
ah @silkenelson I was just noting I don't see the wave8 used in that function...  Maybe my eyes are not parsing well and someone can point out the line.  I see your point about "downstream" possibly being misleading.  My argument here is not so much about what these intensities should be called but rather how they should be represented in the program.  The pervasive elements in any hutch will be that we will always have two intensity measurements: one from something upstream and one from scattering... so program should not really care if its a gasdet or wave8 or cspad or rayonix.  Oh!  "Scattering!" How about that instead of downstream? :D

Reply:
I like scattering. if the scientists have a better name, we can change that later.